### PR TITLE
Make CoalescingBufferQueue test properly release all its buffers

### LINKED ISSUE #989: Deadlock in SSLHandler
Hi,
I think I found a bug in the SSLHandler class (please let me know if instead I'm doing something wrong in my usage of the class). 

I'm using netty 3.6.2 final.

Sometimes (in fact, fairly rarely, I can't reproduce it all the time) one of my integration tests hangs, and looking at the stacktraces it seems to be caused by a deadlock in the SSLHandler: 

One thread (which is handling a message received) takes the handshakeLock as part of the unwrap(), and subsequently does a write, which cause, on the same thread, a channel closed event (the channel was probably being closed at this point), that ends up invoking the channelClosed() method on the same handler, which tries to get hold of the pendingUnencryptedWrites lock.

At the same time, another thread (which is doing a write on the same channel), takes the pendingUnencryptedWrites lock as part of the wrap(), and it then tries to get the handshakeLock, causing the deadlock.

Here are the two stacktraces:

<pre>
"myapp-nioW-8" prio=6 tid=0x000000000d39a000 nid=0x12ca4 waiting for monitor entry [0x000000001118e000]
   java.lang.Thread.State: BLOCKED (on object monitor)
    at org.jboss.netty.handler.ssl.SslHandler.channelClosed(SslHandler.java:1554)
    - waiting to lock &lt;0x00000000ca9d7240&gt; (a java.util.LinkedList)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:88)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:787)
    at org.jboss.netty.channel.SimpleChannelHandler.channelClosed(SimpleChannelHandler.java:216)
    at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:106)
    at com.xxxxx.myapp.stats.ByteStatisticsHandler.handleUpstream(ByteStatisticsHandler.java:63)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:555)
    at org.jboss.netty.channel.Channels.fireChannelClosed(Channels.java:468)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.close(AbstractNioWorker.java:351)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.write0(AbstractNioWorker.java:254)
    - locked &lt;0x00000000ca99c0b0&gt; (a java.lang.Object)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.writeFromUserCode(AbstractNioWorker.java:145)
    at org.jboss.netty.channel.socket.nio.NioServerSocketPipelineSink.handleAcceptedSocket(NioServerSocketPipelineSink.java:99)
    at org.jboss.netty.channel.socket.nio.NioServerSocketPipelineSink.eventSunk(NioServerSocketPipelineSink.java:36)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:775)
    at org.jboss.netty.channel.SimpleChannelHandler.writeRequested(SimpleChannelHandler.java:292)
    at com.xxxxx.myapp.stats.ByteStatisticsHandler.writeRequested(ByteStatisticsHandler.java:120)
    at org.jboss.netty.channel.SimpleChannelHandler.handleDownstream(SimpleChannelHandler.java:254)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:587)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:780)
    at org.jboss.netty.channel.Channels.write(Channels.java:725)
    at org.jboss.netty.channel.Channels.write(Channels.java:686)
    at org.jboss.netty.handler.ssl.SslHandler.wrapNonAppData(SslHandler.java:1140)
    at org.jboss.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1229)
    - locked &lt;0x00000000ca99be40&gt; (a java.lang.Object)
    at org.jboss.netty.handler.ssl.SslHandler.decode(SslHandler.java:910)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:425)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:787)
    at org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142)
    at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
    at com.xxxxx.myapp.stats.ByteStatisticsHandler.handleUpstream(ByteStatisticsHandler.java:63)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:555)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
    at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:107)
    at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:312)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:88)
    at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
    at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
    at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
    at java.lang.Thread.run(Thread.java:662)
</pre>

<pre>
"myapp-scheduler" prio=6 tid=0x000000000d148000 nid=0x1361c waiting for monitor entry [0x000000001068f000]
   java.lang.Thread.State: BLOCKED (on object monitor)
    at org.jboss.netty.handler.ssl.SslHandler.wrap(SslHandler.java:958)
    - waiting to lock &lt;0x00000000ca99be40&gt; (a java.lang.Object)
    - locked &lt;0x00000000ca9d7240&gt; (a java.util.LinkedList)
    at org.jboss.netty.handler.ssl.SslHandler.handleDownstream(SslHandler.java:627)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:587)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:780)
    at org.jboss.netty.channel.SimpleChannelHandler.writeRequested(SimpleChannelHandler.java:292)
    at org.jboss.netty.channel.SimpleChannelHandler.handleDownstream(SimpleChannelHandler.java:254)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:587)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:780)
    at org.jboss.netty.channel.Channels.write(Channels.java:725)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.doEncode(OneToOneEncoder.java:71)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:59)
    at com.xxxxx.myapp.marshalling.CodecHandler.handleDownstream(CodecHandler.java:42)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:587)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:780)
    at org.jboss.netty.channel.SimpleChannelHandler.writeRequested(SimpleChannelHandler.java:292)
    at com.xxxxx.myapp.protocol.ProtocolHandler.writeRequested(ProtocolHandler.java:173)
    at org.jboss.netty.channel.SimpleChannelHandler.handleDownstream(SimpleChannelHandler.java:254)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:587)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:578)
    at org.jboss.netty.channel.Channels.write(Channels.java:704)
    at org.jboss.netty.channel.Channels.write(Channels.java:671)
    at org.jboss.netty.channel.AbstractChannel.write(AbstractChannel.java:248)
[...]
</pre>

Please let me know if you need any additional information, or if I'm doing something wrong in my usage of the SSLHandler.

Thanks a lot
Matteo

### PR DESCRIPTION
**Pull request by louiscryan**
Fixes https://github.com/netty/netty/issues/3989

### CODE REVIEW

File: transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java [hunk 1]
```diff
@@ -215,4 +206,11 @@ private void assertQueueSize(int size, boolean isEmpty) {
             assertFalse(writeQueue.isEmpty());
         }
     }
+
+    private String dequeue(int numBytes, ChannelPromise aggregatePromise) {
+        ByteBuf removed = writeQueue.remove(numBytes, aggregatePromise);
+        String result = removed.toString(Charset.defaultCharset());
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java]: Do we want to use CharsetUtils.US_ASCII ?

File: transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java [hunk 2]
```diff
@@ -67,14 +65,7 @@ public void operationComplete(ChannelFuture future) throws Exception {
 
         cat = Unpooled.wrappedBuffer("cat".getBytes(Charset.defaultCharset()));
         mouse = Unpooled.wrappedBuffer("mouse".getBytes(Charset.defaultCharset()));
-        empty = Unpooled.buffer(0, 1);
-    }
 
-    @After
-    public void tearDown() {
-        ReferenceCountUtil.safeRelease(cat);
-        ReferenceCountUtil.safeRelease(mouse);
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java]: Do we still need to release cat and mouse ?

Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java]: +1. I think we do.

**Comment by Scottmitch** [comment]
@louiscryan - For consistency could we just use the new `dequeue` for the EMPTY buffer checks too?

```java
assertEquals(0, writeQueue.remove(Integer.MAX_VALUE, aggregatePromise).readableBytes());
```

change to 

```java
assertEquals("", dequeue(Integer.MAX_VALUE, aggregatePromise));
```

File: transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java (hunk 2 above)
Comment (louiscryan) [inline transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java]: Yes, sorry I thought I was releasing them in all the tests but that is not the case for the 'empty' tests.

File: transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java (hunk 1 above)
Comment (louiscryan) [inline transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java]: Sure, done

**Comment by louiscryan** [comment]
Adjusted the empty tests as suggested

**Comment by louiscryan** [comment]
PTAL

File: transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java [hunk 3]
```diff
@@ -65,16 +62,13 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
         voidPromise = channel.voidPromise();
 
-        cat = Unpooled.wrappedBuffer("cat".getBytes(Charset.defaultCharset()));
-        mouse = Unpooled.wrappedBuffer("mouse".getBytes(Charset.defaultCharset()));
-        empty = Unpooled.buffer(0, 1);
+        cat = Unpooled.wrappedBuffer("cat".getBytes(CharsetUtil.US_ASCII));
+        mouse = Unpooled.wrappedBuffer("mouse".getBytes(CharsetUtil.US_ASCII));
     }
 
-    @After
```
Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java]: Add this back.

**Comment by Scottmitch** [comment]
@louiscryan - I think there is still 1 more case on line 158.

File: transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java (hunk 3 above)
Comment (louiscryan) [inline transport/src/test/java/io/netty/channel/CoalescingBufferQueueTest.java]: Doh

**Comment by louiscryan** [comment]
Done

**Comment by louiscryan** [comment]
Should be ripe for picking now. Fixed commit message

**Comment by normanmaurer** [comment]
@louiscryan thanks!

Cherry-picked into 4.1 (e6ed3ce6b1bc7424d76a6a8a534592d7ec2ca731) and master (f608a139cfdb0a62ba84d5b006c6351d92578083)

**Pull request closed without merging**